### PR TITLE
NEEDS VOLUNTEER: Communities ported to Protocol Buffers

### PR DESCRIPTION
**Pull request by qstokkink**
Related to #706, #2106.

---

Ported the **AllChannel**, **Bartercast4**, **Channel**, **Demers** and **Template** communities to Google's Protocol Buffers serialization. This implementation is backward compatible with the old wire format (± 2000 LOC will be thrown out when backward compatibility is removed).

Current status:

<table>
<tr>
<th>
TODO
</th>
<th>
Status
</th>
</tr>
<tr>
<td>
Windows dependencies
</td>
<td>
Waiting for verification of installation instructions. <b>NEEDS VOLUNTEER</b>
</td>
</tr>
<tr>
<td>
Code review
</td>
<td>
Requires one or more masochists to actually review 10k lines of code. <b>NEEDS VOLUNTEER</b>
</td>
</tr>
</table>

### CODE REVIEW

**Comment by lfdversluis** [comment]
add to whitelist

**Comment by whirm** [comment]
add to whitelist

**Comment by qstokkink** [comment]
Oh right, the testing framework does not have the Protocol Buffers python package installed. What would be the proper place to add this?

**Comment by whirm** [comment]
@qstokkink `debian/control` needs to be updated. I'll update the Ansible scripts.

**Comment by whirm** [comment]
bbq has protobuf now

**Comment by qstokkink** [comment]
@whirm thanks!

**Comment by qstokkink** [comment]
retest this please

**Comment by whirm** [comment]
Let me know when you need the allchannel experiment to be enabled. I disabled it for now as it still hasn't the free cluster discovery stuff I made integrated.

**Comment by qstokkink** [comment]
@whirm 
__tl;dr:__ I think now would be a good time to enable it. 

__long answer:__
All tests on the backward compatibility code pass (on Linux) and all but 6 tests for the pure new code pass (which -seem to- fail because AllChannel can't find anything in a live search using only the new protocol). At this point I will start writing my own tests to validate new behavior and interactions.

So, I think now would be a good time to see what kind of performance benefit would be gained from the new system. Or in the worst case: what kind of performance loss. Then it can be decided if this project is worth merging.

**Comment by whirm** [comment]
Oh, apparently I didn't disable it for dispersy, you just need for the rest to pass for the experiment to be triggered.

**Comment by synctext** [comment]
Nice progress on this PR. Thousands of lines of code altered... wow.

**Comment by qstokkink** [comment]
For my own reference, this is the passing build using the old protocol (the backward compatibility switch is working as intended):
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_linux/2201/
I will focus my attention on passing the other tests for the new protocol a.s.a.p.

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
Alright the results are in for the old protocol (with new interface):
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_linux/2380/
And the new Protocol Buffers implementation (same interface), using the old *create missing channel* (Channel community) and *create channel search* (AllChannel community) functions:
https://jenkins.tribler.org/job/GH_Tribler_PR_tests_linux/2389/

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
@devos50 How much of the communities are you already covering with the REST API tests? It seems you are already covering most (if not all) of the allchannel and channel communities. 
This would put the focus of my tests more on the compatibility interactions.

**Comment by qstokkink** [comment]
retest this please

**Comment by devos50** [comment]
@qstokkink it's not that good, however, the core functionality of most communities (search, all channel, channel, tunnel) are covered with our (GUI) tests.

I think we still need some stable and small unit tests to test the logic within these communities without being dependent on a running Tribler session but it's not really a top priority issue.

**Comment by qstokkink** [comment]
@devos50 It's really hard (read: ugly private-field-dependant code) to separate the Session from a community. For instance the AllChannel community will either make a database stub or a real handler, internally, depending on whether or not it is *initialized* with a session. To make matters worse, these stubs are not shared between communities. Furthermore, because of the coupling side effects of the real handlers, checking proper database calls is only really possible using a Session's created databases.

That said, using a very limited SessionStartupConfig it is possible to make some somewhat nimble unittests. What I did for my AllChannel unittests (local WIP, keep an eye out for it soon), is generation of the databases using a Session and then overwrite the stubs (private field access <- not cool) with the proper databases. This is the only way to get everything working with Dispersy's DebugNodes without getting Session singleton errors.

Since your unittests are initialized and checked more elegantly than mine, I would have dropped my approach in favor of yours. However, since there is still some merit to it apparently, I will continue work on the low-level unittests.

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
For anyone interested: the first results show an average 14% reduction in test completion time using Protocol Buffers. These results are based on the time it takes between the AllChannel community to create a message on one node and handle it on another node (ergo mostly serialization and unserialization of messages).

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
@whirm I created [build instructions for (x64) Windows, which are not quite as elegant as I would have liked](https://github.com/qstokkink/tribler/blob/protobufserialization_rc/doc/development/development_on_windows.rst#protocol-buffers). I also see I triggered 5000 builds while wrestling with the ReStructuredText syntax.

File: doc/development/development_on_windows.rst [hunk 1]
```diff
@@ -110,6 +110,46 @@ VLC
...
+    "%LOCALAPPDATA%\Programs\Common\Microsoft\Visual C++ for Python\9.0\vcvarsall.bat"
+    "%LOCALAPPDATA%\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\bin\amd64\vcbuild.exe" /upgrade libprotobuf.vcproj
+    "%LOCALAPPDATA%\Programs\Common\Microsoft\Visual C++ for Python\9.0\VC\bin\amd64\vcbuild.exe" /useenv libprotobuf.vcproj "Release|Win32"
+
+If everything completed correctly this should have created the file:
+:code:`%FULLSOURCELOCATION%\vsprojects\Release\libprotobuf.lib`
+Copy **and rename** this file to:
+:code:`%FULLSOURCELOCATION%\python\protobuf.lib`
+
+Finally run :code:`python %FULLSOURCELOCATION%\python\setup.py install --cpp_implementation` **with a 32-bit python version**.
```
Comment (whirm) [inline doc/development/development_on_windows.rst]: We support both 32 and 64 bit builds of Tribler

Comment (qstokkink) [inline doc/development/development_on_windows.rst]: @whirm The most recent version has the 64 bit instructions.

**Comment by qstokkink** [comment]
Could someone with a Mac try the following for me? @devos50 are you available?
<hr>
Experimental/unverified Mac instructions:

1. Download the latest version from https://developers.google.com/protocol-buffers/docs/downloads
2. Execute `make install` in the top folder.
3. Execute `python setup.py build --cpp_implementation` in the `\python` folder.

Test:

1. Execute `python -c "import google.protobuf.pyext._message"`: this should not generate any import errors.

**Comment by devos50** [comment]
@qstokkink no need to compile it by yourself :)

```
brew install protobuf
```

```
MBP-van-Martijn:tribler martijndevos$ python -c "import google.protobuf.pyext._message"
MBP-van-Martijn:tribler martijndevos$
```

**Comment by qstokkink** [comment]
@devos50 thanks! That's much easier than the Windows build.

**Comment by qstokkink** [comment]
@whirm Could you perform the Mac `brew install protobuf` installation?
If it's ok with you, I can come in monday to perform the Windows 32 & 64 bit installations.

**Comment by lfdversluis** [comment]
Just asked @devos50 to install it, will be on it soon!

**Comment by devos50** [comment]
@qstokkink installed

**Comment by lfdversluis** [comment]
retest this please - Let's see

**Comment by qstokkink** [comment]
Big whoop. :D

**Comment by qstokkink** [comment]
For reference, the new testing coverage:
<table>
<tr>
<th>
Coverage:
</th>
<th>
Line
</th>
<th>
Statement
</th>
</tr>
<tr>
<td>
AllChannel (Old)
</td>
<td>
57%
</td>
<td>
38%
</td>
</tr>
<tr>
<td>
AllChannel (New)
</td>
<td>
68% <b>(+11%)</b>
</td>
<td>
51% <b>(+13%)</b>
</td>
</tr>
<tr>
<td>
Channel (Old)
</td>
<td>
44%
</td>
<td>
26%
</td>
</tr>
<tr>
<td>
Channel (New)
</td>
<td>
69% <b>(+25%)</b>
</td>
<td>
50% <b>(+24%)</b>
</td>
</tr>
</table>

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
@devos50 or @lfdversluis could one of you verify [the build instructions for Windows](https://github.com/qstokkink/tribler/blob/e896577b19728e909f15e670162f6b6393785825/doc/development/development_on_windows.rst#protocol-buffers) (on a Proxmox server). I can't verify the instructions, because I wrote the instructions.

**Comment by devos50** [comment]
@qstokkink protobuf is available on pypi: https://pypi.python.org/pypi/protobuf/3.0.0b4

So isn't 'pip install protobuf' sufficient?

**Comment by qstokkink** [comment]
@devos50 Already tried that, it doesn't ship with the required headers.

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
Here are some side-by-side results, for anyone interested.
<table>
  <tr>
    <th>Graph</th>
    <th>Old</th>
    <th>New<br></th>
  </tr>
  <tr>
    <td>Send</td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>Received<br></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>UTimes<br></td>
    <td></td>
    <td></td>
  </tr>
  <tr>
    <td>STimes<br></td>
    <td></td>
    <td></td>
  </tr>
</table>
For now all _correctly handled messages_ [are being registered as dropped due to lack of proof](https://jenkins.tribler.org/job/pers/job/clone_gumby_qstokkink/101/artifact/output-new/dispersy-dropped-msg-distribution.txt/*view*/), even though they are authorized and are being handled properly. Once this has been fixed, this PR will be truly done.

**Comment by synctext** [comment]
fascinating changes. In this first picture, the PrimeSource sends out 2700 or so messages out. Within the new situation that injector only sends 2400 or so.

**Comment by synctext** [comment]
Mean bug, timeout for the missing-proof. Over 1 week spend investigating, but no insight yet or fix in sight.

**Comment by qstokkink** [comment]
Ok, I **FINALLY** figured out what is causing the massive drop count, [it was actually in the comments all along](https://github.com/Tribler/tribler/blob/devel/Tribler/community/channel/community.py#L104). The _Torrent_ messages are not processed in priority order, causing them to drop (which is intended behavior if they had the same priority).

**Comment by synctext** [comment]
congrats!
clearly we should document the stress and drop behavior...

**Comment by synctext** [comment]
can you please fix & improve the degraded behavior to be more intuitive?

**Comment by qstokkink** [comment]
@synctext I guess the intuitive solution would be to declare the message types that need to be processed before it, per message type.
So instead of declaring an arbitrary priority value per message, one would specify a list of other message types which should be processed first if they exist in the processing queue.

I don't think there is a way to do this completely automated, but I am open to suggestions.

**Comment by qstokkink** [comment]
This is a public MEMO: the 27 new unit tests for the AllChannel and Channel communities should be extracted from this PR into another PR.

**Comment by qstokkink** [comment]
retest this please

**Comment by qstokkink** [comment]
@devos50 Any idea why all of my PR tests are being aborted by anonymous?
```
...
[TemplateProject] Starting builders from: Test_tribler_devel
Build was aborted
Aborted by anonymous
Archiving artifacts
...
```

**Comment by devos50** [comment]
@qstokkink probably because if the spelling checker fails, it aborts the execution of the whole job. You should check if that's the case and if so, fix it so the failure gets ignored and the job continues to execute.

**Comment by qstokkink** [comment]
@devos50 thanks, forgot a checkbox in GH_Tribler_PR :)
retest this please

**Comment by qstokkink** [comment]
Well now.. I appear to be a bit of a ding dong. The reason the graph wasn't showing any records being received is because I didn't tell it to plot "basemsg" messages. 😕 

Anyway, there are no known bugs remaining for this PR now (/again).

**Comment by devos50** [comment]
@qstokkink can you explain this error? Should I reinstall protobuf on the Windows machines?

```
Error Message

cannot import name _message
-------------------- >> begin captured logging << --------------------
DEBUG   1484870014         _legacy:twisted:154  PythonLoggingObserver hooked up
--------------------- >> end captured logging << ---------------------
Stacktrace

  File "C:\Python27\lib\unittest\case.py", line 329, in run
    testMethod()
  File "C:\Python27\lib\site-packages\nose\loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "C:\Python27\lib\site-packages\nose\importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "C:\Python27\lib\site-packages\nose\importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\Test\Core\Modules\RestApi\test_util.py", line 4, in <module>
    from Tribler.Core.Modules.restapi.util import convert_search_torrent_to_json, convert_db_channel_to_json, \
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\Core\Modules\restapi\util.py", line 12, in <module>
    from Tribler.community.channel.community import ChannelCommunity
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\community\channel\community.py", line 11, in <module>
    from Tribler.community.basecommunity import BaseCommunity
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\community\basecommunity.py", line 3, in <module>
    from Tribler.community.TriblerProtobufSerialization.serializer import Serializer
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\community\TriblerProtobufSerialization\serializer.py", line 10, in <module>
    import messages
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\community\TriblerProtobufSerialization\messages\__init__.py", line 4, in <module>
    MESSAGES = [__import_handle('allchannel_pb2'),
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\community\TriblerProtobufSerialization\messages\__init__.py", line 2, in __import_handle
    return __import__(name, globals(), locals(), [], -1)
  File "C:\workspace\jenkins\workspace\GH_Tribler_PR_tests_win32\tribler\Tribler\community\TriblerProtobufSerialization\messages\allchannel_pb2.py", line 6, in <module>
    from google.protobuf import descriptor as _descriptor
  File "C:\Python27\lib\site-packages\google\protobuf\descriptor.py", line 50, in <module>
    from google.protobuf.pyext import _message
'cannot import name _message\n-------------------- >> begin captured logging << --------------------\nDEBUG   1484870014         _legacy:twisted:154  PythonLoggingObserver hooked up\n--------------------- >> end captured logging << ---------------------'
```

**Comment by devos50** [comment]
@qstokkink this breaks backwards compatibility right? If so, I would save this PR for Tribler 7.1 or maybe later.

**Comment by qstokkink** [comment]
@devos50 (1) It has never been installed on Windows (I still need someone to verify the Windows install instructions). 

(2) In its current form the code IS backwards compatible (by means of additional code, which can be removed when backwards compatibility is broken); that said, I think this would be a good one to postpone after the 7.0 release, yes.

**Comment by qstokkink** [comment]
Closing this in favor of performing a clean protobuf implementation on top of IPv8. Additional unit tests have already been exported to another PR (#2911).
